sql/parser: remove an assymetry

knz opened the pull request:
First 4 commits from #27214 and priors.

Earlier I made the choice to extend CockroachDB so that it would allow
the array indexing syntax `[...]` for more expressions forms than
PostgreSQL: pg only allows array indexing after parentheses, and
CockroachDB now supports it after any "closed form" expression (c_expr
in pg, d_expr in CockroachDB).

(Context: a closed form expression is one for which it is always
unambiguous where is the end of the syntax, e.g. using parentheses, a
final keyword, etc. A closed form non-terminal X can be used in a
grammar like "E := X | X op X" without any ambiguity as to whether
"op" belongs inside X or as part of E.)

Conceptually, the array indexing grammar should be part of c_expr in
both pg and crdb, because it is itself closed form: a (closed form)
expression followed by `[...]` is also closed form (up to the `]`).
However, this creates a lookahead-based ambiguity due to SQL's
`typename SCONST` rule, which the LALR generator is unable to
resolve. So I chose to resolve this by splitting c_expr which
recognizes the array indexing from d_expr which recognizes every other
closed form expression.

However in doing so I made a mistake: d_expr should contain all closed
form expressions, including CASE..END and EXISTS(...) -- I did not
include them in d_expr and placed them in c_expr instead. There was no
good reason for doing so.

Release note (sql change): CockroachDB now allows clients to use the
array indexing operator `[...]` after CASE expressions.

knz committed:
sql: make AS OF SYSTEM TIME accept any constant expression

Prior to this patch, AS OF SYSTEM TIME relied on the grammar to
restrict its operand to just constants. This was an unfortunate
approach because it prevented other constant expressions like `('a' ||
'b')::INTERVAL` (or, encountered during tests, `'1ms':::INTERVAL`).

This patch corrects this situation by relaxing the grammar constraint
and instead verifying the const-ness of the expression after type
checking.

Release note (sql change): AS OF SYSTEM TIME can now use some more
complex expressions to compute the desired time stamp.

knz committed:
sql: fix EXPERIMENTAL CONFIGURE ZONE

Two fixes where:

- fix reporting of ALTER ... CONFIGURE ZONE (was broken)
- relax constraint on operand

First fix: an implementation of `WalkExpr` was missing.

Second fix:
The implementation of EXPERIMENTAL CONFIGURE
ZONE (`sql.SetZoneConfig`) was already perfectly capable to process
arbitrarily complex expressions for the YAML configuration string.

Somehow however the grammar had a restriction to only accept constant
expressions. This restriction is not needed, so this patch removes it.

Release note (sql change): the ALTER ... EXPERIMENTAL CONFIGURE ZONE
statement now accepts arbitrary scalar expressions (including possibly
containing sub-queries) to compute the YAML operand.

Release note (bug fix): ALTER ... EXPERIMENTAL CONFIGURE ZONE is
now properly tracked in statement statistics.

knz committed:
sql/parser: remove the now-unused a_expr_const non-terminal

Release note: None

knz committed:
sql/parser: remove an assymmetry

Earlier I made the choice to extend CockroachDB so that it would allow
the array indexing syntax `[...]` for more expressions forms than
PostgreSQL: pg only allows array indexing after parentheses, and
CockroachDB now supports it after any "closed form" expression (c_expr
in pg, d_expr in CockroachDB).

(Context: a closed form expression is one for which it is always
unambiguous where is the end of the syntax, e.g. using parentheses, a
final keyword, etc. A closed form non-terminal X can be used in a
grammar like "E := X | X op X" without any ambiguity as to whether
"op" belongs inside X or as part of E.)

Conceptually, the array indexing grammar should be part of c_expr in
both pg and crdb, because it is itself closed form: a (closed form)
expression followed by `[...]` is also closed form (up to the `]`).
However, this creates a lookahead-based ambiguity due to SQL's
`typename SCONST` rule, which the LALR generator is unable to
resolve. So I chose to resolve this by splitting c_expr which
recognizes the array indexing from d_expr which recognizes every other
closed form expression.

However in doing so I made a mistake: d_expr should contain all closed
form expressions, including CASE..END and EXISTS(...) -- I did not
include them in d_expr and placed them in c_expr instead. There was no
good reason for doing so.

Release note (sql change): CockroachDB now allows clients to use the
array indexing operator `[...]` after CASE expressions.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/27210)
<!-- Reviewable:end -->

knz committed:
server: clarify TestReportUsage

I need this to investigate a test failure in an unrelated PR.

Release note: None

justinj reviewed (commented):
Code LGTM, are we sure being less restrictive is a good thing though? Seems to me like we might be setting ourselves up for unforeseen grammar incompatibilities.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/27210#-:-LGzbACu5GrTbVzW4psj:b30h96q)* status: :shipit: complete! 0 of 0 LGTMs obtained



<!-- Sent from Reviewable.io -->

knz commented:
yeah you're right, and I am changing my mind as a result. Let's let this change sleep for now.
